Add option to change SVE vector length for current and children processes

SwapnilGaikwad opened the pull request:
Current coreclr assumes SVE vector length as 128 bits leading it to limit size of Vector<T> to 16 bytes.
While executing on a platform offerring higher vector lengths, such as 256 bits, lead to using registers of larger size.
This leads to incorrect results, e.g., while using unzip even (uzp1 instruction).
Here C# expects processing based on half of the vectors (size 128bits) while the actual result is based on full vectors (size 256bits).


Add `DOTNET_MaxVectorLength=N` flag where N is desired SVE vector length in bytes for the current execution.
Let M is the max/current vector length and N is the vector length specified with `DOTNET_MaxVectorLength` option, then V is the new vector length for the current execution.
- If N < M, N % 16 == 0 (a valid vector length), V = N
- If N < M, N % 16 != 0 (an invalid vector length), V = M
- If N > M (N can be a valid or invalid length), V = M

SwapnilGaikwad committed:
Add option to change SVE vector length for current and children processes.

SwapnilGaikwad commented:
This patch is required for #101294 while running on a system that offers SVE vector length greater than 128.

a74nh commented:
This PR fixes the issues I've been seeing implementing `AddAcross`.

We've both been using V1 machines. On an N2 this PR is not required, and won't cause any effects.

tannergooding commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -790,6 +790,10 @@ RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableArm64Sha256,            W("EnableArm64Sh
 RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableArm64Sve,               W("EnableArm64Sve"),            1, "Allows Arm64 SVE hardware intrinsics to be disabled")
+#if defined(TARGET_LINUX) && (defined(DEBUG) || defined(_DEBUG))
+// TODO-SVE: Once coreclr supports vector lengths >16bytes, the default should be -1.
+RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_MaxVectorLength,           W("MaxVectorLength"),          16, "Specifies maximum size of the vector length in bytes while using SVE on Arm64", CLRConfig::LookupOptions::ParseIntegerAsBase10)
```
Why can't we just use `MaxVectorTBitWidth` for this?

tannergooding replied:
Right, I understand what this PR is trying to fix.

What I'm saying is that we already have an option to control the maximum size of `Vector<T>` and that is `DOTNET_MaxVectorTBitWidth`, so why aren't we just using that to influence whether or not `prctl` is called instead?

Arm64 should really just have a bit of code that looks vaguely like this: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/codeman.cpp#L1274-L1287, where it:
1. Queries the current SVE vector length
2. Compares that against `MaxVectorTBitWidth`
3. If the user has explicitly set a maximum width and its less than the current bit width, we then have the option of either disabling SVE (undesirable) or using the OS functionality (such as `prctl`) to match the user targeted size
4. In the default experience where the user hasn't explicitly set a maximum width (it is `0`), then long term we should do nothing but short term we should use the OS functionality to fix the size to 128-bits, until the support for larger widths can be brought online

SwapnilGaikwad replied:
Makes sense, I'll move this code to use `MaxVectorTBitWidth`. Thanks for this suggestion 👍

kunalspathak replied:
@SwapnilGaikwad - are you planning to push this patch as well as remove it from `#ifdef DEBUG` in `codeman.cpp`?

SwapnilGaikwad replied:
Yup, removed the changes from `#ifdef DEBUG` as they now relay on `MaxVectorTBitWidth`.

tannergooding commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1525,6 +1528,18 @@ void EEJitManager::SetCpuInfo()
 
     if (((cpuFeatures & ARM64IntrinsicConstants_Sve) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableArm64Sve))
     {
+#if defined(TARGET_LINUX) && (defined(DEBUG) || defined(_DEBUG))
```
Why is this restricted to DEBUG only?

I also expect that we need a path for Windows and should likely treat SVE as unsupported if the vector length is larger and restricting the size (via `prctl` or equivalent on other OS) fails.

SwapnilGaikwad replied:
This option is added for development purpose only.  It's primarily added to aid implementation of API, and its testing on a 256bit SVE enabled V1 system that we have access to. When Vector<T> would use the entire vector length, it may become redundant and be removed.

tannergooding replied:
I don't think it being for development purposes only is a good thing long term and it probably doesn't match up with the intent of `DOTNET_MaxVectorTBitWidth`.

We don't really want to disable SVE on hardware with 256-bit vectors just because a user has said they want 128-bit vectors, using the OS feature like `prctl` should be a much better option so that they can still get SVE usage in that scenario.

kunalspathak replied:
> I also expect that we need a path for Windows and should likely treat SVE as unsupported if the vector length is larger and restricting the size (via `prctl` or equivalent on other OS) fails.

I will check with Windows team on what is the equivalent API.

JasonLinMS replied:
Kunal asked me to share how this works in Windows (I developed the SVE support in Windows for the next release). We felt that being able to dynamically change the vector length at runtime would be messy, since there's no save/restore mechanism for these vector length changes, hence you could get into situations where for example some code calls into a library, the library decreases the VL for some reason, but then has a bug where under certain conditions it fails to restore the vector length, so after the caller gains control again it would be running with a decreased vector length indefinitely, etc.

We do recognize the need to change the vector length for various purposes (testing, perf, compat, etc), so we did add a CreateProcess parameter that allows the vector length to be specified during process creation, but after a process has been created, its vector length cannot be changed. There's also registry keys that can be set to change the vector length on a per-process basis (IFEO settings), or for all processes in the system, but these registry-based methods are typically used for development and testing only.

By default, we'll use the highest VL supported by the system and the underlying hypervisor. So the above mechanisms will only be necessary when you want to decrease the VL for processes.

Happy to explore this topic more with you.

tannergooding replied:
Thanks for the context @JasonLinMS. 

I definitely agree that changing this during the lifetime of the process is messy/problematic and the intent in general is to not allow that for .NET either.

I think it's probably worth us having a short meeting (.NET, Windows, and Arm) to see if we can discuss things here and see if we can find something that generally works for everyone. CC. @jkotas 

-----------------

In the case of .NET, we would really only need the capability to set this once as part of our own startup before any user code has executed. We don't have any intent to change this dynamically (although there is the SME feature that makes this a little bit more complicated) and the ideal for user code is to write size agnostic algorithms so that it doesn't matter what size the hardware actually supports.

However, the official Arm64 SVE/Vector ABIs (https://github.com/ARM-software/abi-aa?tab=readme-ov-file#abi-for-the-arm-64-bit-architecture-with-sve-support) do define the ability to say a given API expects a particular size and there may be cases where a user needs to fix it themselves (potentially just for testing purposes or giving users a workaround for a bug). As such, the ability for a process to set the size for itself is beneficial, especially if that can be shared across all Arm64 capable operating systems.

JasonLinMS replied:
Yes, a meeting to discuss this sounds good.

kunalspathak replied:
scheduled.

kunalspathak replied:
Once other feedback is addressed, this change can go in without having us to wait for Windows support. 

I think this should also check the other requirements such as it should by 128-bit increments only and maximum should be 2048 and such.

a74nh replied:
> we did add a CreateProcess parameter that allows the vector length to be specified during process creation, but after a process has been created, its vector length cannot be changed.

This is a very sensible choice. In Linux, changing the vector length inside a library and not restoring it before returning is generally seen as undefined behaviour. OpenJDK doesn't want to trust that, and after calling external routines it inserts checks to confirm the vector length hasn't changed. Neither of which is ideal.

> However, the official Arm64 SVE/Vector ABIs (https://github.com/ARM-software/abi-aa?tab=readme-ov-file#abi-for-the-arm-64-bit-architecture-with-sve-support) do define the ability to say a given API expects a particular size

The PCS isn't clear if the vector length must remain fixed. I've raised a bug against it [here](https://github.com/ARM-software/abi-aa/issues/260). It would be good to have a clear statement.



For coreclr, if changing the vector length is only for debugging then a wrapper script/binary that just launches coreclr with the correct vector length might be good enough. That would work for Windows and Linux. Ideally I'd still like to keep this PRs mechanism for use in Linux.

SwapnilGaikwad committed:
Use maxVectorTBitWidth to get desired SVE length

SwapnilGaikwad commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -42,6 +45,13 @@
 #endif
 
+#if defined(TARGET_LINUX) && !defined(PR_SVE_GET_VL)
+#define PR_SVE_SET_VL           50	/* set task vector length */
+#define PR_SVE_GET_VL           51	/* get task vector length */
+#define PR_SVE_VL_LEN_MASK      0xffff
+#define PR_SVE_VL_INHERIT       (1 << 17) /* inherit across exec */
```
Defining these vars explicitly as there are some builds failing, such as [this](https://dev.azure.com/dnceng-public/public/_build/results?buildId=649553&view=logs&j=e2af9a1a-ae86-5313-2146-b7306ad4db64&t=8f10af5c-d1dd-5c33-de37-3925c5c2a252&l=2740).
Kindly suggest if there is a better way of doing this.

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1525,6 +1535,22 @@ void EEJitManager::SetCpuInfo()
+        int maxVectorLength = (maxVectorTBitWidth >> 3);
+
+        // Limit the SVE vector length to 'maxVectorLength' if the underlying hardware offers longer vectors.
+        if ((prctl(PR_SVE_GET_VL, 0,0,0,0) & PR_SVE_VL_LEN_MASK) > maxVectorLength)
```
This will set the SVE vector length only for the current thread only. It won't set it for threads that has been started already. Is that correct?

How does it work for new threads? Do new threads inherit SVE vector length of the parent thread or do new threads inherit SVE vector length of the process?

Are we sure that none of the libraries that has been initialized by this point have not cached the vector length?

https://www.man7.org/linux/man-pages/man2/prctl.2.html has explicit warning about use of this API only if you know what you are doing.

It seems that allowing the SVE vector length to be set only before process start is the only reliable option. Setting it here may break all sorts of random things. I am not convinced that we know what we are doing by setting it here.

tannergooding replied:
Do you mean in the case of something like the C runtime or in the case of something else hosting the CLR?

Are you thinking the only valid thing for us to do here is to fail to launch for an SVE mismatch (AOT) and to just disable SVE usage otherwise (JIT)?

jkotas replied:
Right, this is obviously not compatible with hosting. (We have no reliable way to tell whether we are hosted.) 

Even without hosting and external libraries outside our control in the picture, there are number of our own threads created in the process by this point (PAL, EventPipe) that will have a wrong size. It is hard to guarantee that none of these threads is going to wonder into managed code.

> Are you thinking the only valid thing for us to do here is to fail to launch for an SVE mismatch (AOT)

I do not see a better option. It suggests that the design we are working with is questionable since it does not work well for AOT.

> just disable SVE usage otherwise (JIT)

Yes, if the JIT is not able to accommodate the SVE length that the process was started with.

tannergooding replied:
> If this is only used for debugging/testing and never used in production, then I lean towards it's fine. Anything more then maybe not.

In general these switches are primarily there for debugging/testing purposes. However, they also generally exist as a way to disable or limit intrinsic support if a library is found to have a blocking bug.

It's not great that we can't help setup the process to achieve success, but its also not the end of the world and is something we can ideally give user guidance around.

> Or should DOTNET_MaxVectorTBitWidth be X86 only?

I think it's fine for us to respect it still, that's functionally what AOT compiled for a particular SVE size would have to do after all. It's just a different way to disable SVE support.

jkotas replied:
The documented switches have to be reliable. `DOTNET_MaxVectorTBitWidth` is documented switch. 

It does not sound like that this switch can be reliable. It means that it should have different name, and ideally be a debug-only switch. We do not want to be dealing with inscrutable crashes caused the different parts of the process being configured to different vector sizes.

tannergooding replied:
Right. I should have clarified I meant that given your input that we shouldn't take this PR to call `prcrtl` because its unreliable, that the alternative where we simply just don't use SVE if its larger than the `DOTNET_MaxVectorTSize` is fine and still inline with the currently documented behavior for that config switch.

If we provided anything around `prctl` (and it sounds like we're leaning towards `no`), it would need to be a separate undocumented switch, potentially debug only. -- I don't think we have the need to add that given our current testing needs and the known sizes (128 and 256) we'll want to support for existing SVE capable hardware (both consumer and server/cloud).

kunalspathak replied:
Thanks @tannergooding for clarifying some of the things offline, so it eventually boils down to:

 ```c#
 int VectorTLength = 0;
 int SystemVectorTLength = Get_System_VL();
 
 if (DOTNET_MaxVectorTBitWidth == 0)
 {
    // If we fix getVectorTByteLength() to return system length - then use system provided length
    // Otherwise use 128
    VectorTLength = SystemVectorTLength;
 }
 else if (DOTNET_MaxVectorTBitWidth >= SystemVectorTLength)
 { 
    // For a 256-bit machine, if user provides DOTNET_MaxVectorTBitWidth=512, we will use the
    // maximum available length of 256-bits
    VectorTLength = SystemVectorTLength;
 }
 else if (DOTNET_MaxVectorTBitWidth < SystemVectorTLength)
 {
    // For a 256-bit machine, if user provides DOTNET_MaxVectorTBitWidth=128, we do not want
    // to update it using syscall because that has implications on already initialized components
    // as it might have stale vector length. In that case, disable SVE.
    //
    // If user really wants to downgrade the size, they can call `prctl` or windows `CreateProcess()` 
    // to limit the size before launching dotnet process
    VectorTLength = 128
    DisableSve();
 }
 ```

To implement `Get_System_VL()`, we can use `prctl PR_SVE_GET_VL`, it will be good to use `CNTB` or an equivalent instruction because that way, it will be OS agnostic. We will need that anyway for `getVectorTByteLength()` when `DOTNET_MaxVectorTBitWidth` is not set.

With that said, there is no need to introduce a different environment variable that is `DEBUG` only to support the downgrade scenario.

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Use CNTB to determine current vector length

kunalspathak commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1525,7 +1525,16 @@ void EEJitManager::SetCpuInfo()
+        {
+            // Enable SVE only when user specified vector length larger than or equal to the system
+            // vector length. When eabled, SVE would use full vector length available to the process.
+            // For a 256-bit machine, if user provides DOTNET_MaxVectorTBitWidth=128, disable SVE.
```
When `DOTNET_MaxVectorTBitWidth` is not specified, it is `0` and in that case, we should use the full vector length, the system offers (as the comment says). In that case, we should set `CPUCompileFlags.Clear(InstructionSet_VectorT256);` on 256-bit machine and `CPUCompileFlags.Clear(InstructionSet_VectorT256);`  and `CPUCompileFlags.Clear(InstructionSet_VectorT512);`  on a 512-bit machine.

tannergooding replied:
We wouldn't want to `CPUCompileFlags.Clear`, as that's implicit

Rather instead we would `CPUCompileFlags.Set(InstructionSet_VectorT256);` if the reported SVE length is `256-bits` and `CPUCompileFlags.Set(InstructionSet_VectorT512);` if the reported SVE length is `512-bits`.

They should be off unless otherwise set, but we double check that via a cleanup check anyways (which should be moved to be shared with Arm64): https://github.com/dotnet/runtime/blob/22aa47e02bb28f5fc009f06b7f526ee5aa33dfd7/src/coreclr/vm/codeman.cpp#L1564-L1573

-- I think we also need a TODO here explaining that we're artificially restricting the size to 128-bits for the time being, as the support for larger vector sizes hasn't been plumbed through for Arm64 yet.

kunalspathak replied:
> We wouldn't want to CPUCompileFlags.Clear, as that's implicit

Yes, that's what I meant...if they are already enabled, then we should disable it or vice-versa. 

> I think we also need a TODO here explaining that we're artificially restricting the size to 128-bits for the time being,

Yes, I think that's what we should do and leave the change of setting `Vector256`, etc. for later PR, when we address some of the things in https://github.com/dotnet/runtime/issues/101477 , specifically `getVectorTByteLength()`.

kunalspathak reviewed (changes requested):
We need to set the system vector length if `DOTNET_MaxVectorTBitWidth` is not specified.

kunalspathak commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1257,6 +1257,9 @@ void EEJitManager::SetCpuInfo()
     int cpuFeatures = minipal_getcpufeatures();
 
+    // Get the maximum bitwidth of Vector<T>, rounding down to the nearest multiple of 128-bits
+    uint32_t maxVectorTBitWidth = (CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_MaxVectorTBitWidth) / 128) * 128;
```
ah, somehow i was not looking this correctly. Makes sense.

tannergooding commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1525,7 +1525,16 @@ void EEJitManager::SetCpuInfo()
     if (((cpuFeatures & ARM64IntrinsicConstants_Sve) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableArm64Sve))
     {
-        CPUCompileFlags.Set(InstructionSet_Sve);
+        int maxVectorLength = (maxVectorTBitWidth >> 3);
```
nit, the following is much more readable/understandable and modern compilers already optimize it to `>> 3` since `maxVectorTBitWidth` is unsigned:
```suggestion
        uint32_t maxVectorLength = maxVectorTBitWidth / 8;
```

SwapnilGaikwad replied:
Done

tannergooding commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1525,7 +1525,16 @@ void EEJitManager::SetCpuInfo()
     {
-        CPUCompileFlags.Set(InstructionSet_Sve);
+        int maxVectorLength = (maxVectorTBitWidth >> 3);
+        uint64_t systemVectorTLength = GetSystemVectorLength();
```
nit: This name is potentially confusing since we have the `System` namespace and many concepts related to `Vector`.

Something that explicitly says `Sve` in the name would be much clearer IMO. Perhaps `GetSveLengthFromOS()` or `GetSveLengthFromSystem()` instead, just to help avoid ambiguities here?

SwapnilGaikwad replied:
Done

tannergooding commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1525,7 +1525,16 @@ void EEJitManager::SetCpuInfo()
+        {
+            // Enable SVE only when user specified vector length larger than or equal to the system
+            // vector length. When eabled, SVE would use full vector length available to the process.
+            // For a 256-bit machine, if user provides DOTNET_MaxVectorTBitWidth=128, disable SVE.
```
This comment is implied by the previous statement. Is it just giving an example? If so, can we explicitly state so, such as:
> For example, if the reported SVE length is 256-bits but the `MaxVectorTBitWidth=128`, then we need to disable SVE support.

SwapnilGaikwad replied:
Put a bit more concise comment now.

tannergooding commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1525,7 +1525,16 @@ void EEJitManager::SetCpuInfo()
+        int maxVectorLength = (maxVectorTBitWidth >> 3);
+        uint64_t systemVectorTLength = GetSystemVectorLength();
+
+        if (maxVectorLength >= systemVectorTLength)
```
It looks like we're missing the handling to artificially restrict `maxVectorLength` to 128 for the time being, which is needed until the handling for `256-bit` or `512-bit` vectors can be plumbed through the JIT on Arm64 (I don't expect that's a lot of work, since its already done for x64, but I expect there to be some Arm64 specific paths that need adjusted)

kunalspathak replied:
> handling to artificially restrict maxVectorLength to 128 for the time being

yes, I am hoping that we will do it in this PR as I mentioned in https://github.com/dotnet/runtime/pull/101295#discussion_r1588105539 and do the handling of 256-bit later as a separate PR.

SwapnilGaikwad replied:
Marking this as resolved. https://github.com/dotnet/runtime/pull/101295#issuecomment-2093312795 describes the new behaviour that also handles the default case (maxVectorTBitWidth = 0).

tannergooding commented on src/coreclr/vm/codeman.h:
```diff
@@ -1912,6 +1912,16 @@ protected :
+#ifdef __GNUC__
+    __attribute__((target("sve")))
+#endif
+    inline UINT64 GetSystemVectorLength()
+    {
+        UINT64 size;
+        __asm__ __volatile__("cntb %0" : "=r"(size));
+        return size;
+    }
```
I'd expect this to fail for MSVC, which doesn't allow inline assembly?

Can we just call `svcntb()` on Windows instead, or is there potentially an official OS API for this?

a74nh replied:
I wasn't sure this was available for MSVC yet. If so, then great, that's easier.

Is there a min version of MSVC needed to build coreclr?

tannergooding replied:
Hmmm, it might not be available yet, at least its not available in the MSVC version CI is using. -- @kunalspathak might know better when that support is expected to land.

I expect we need some ifdef here regardless since `__asm` is explicitly unsupported on MSVC for anything but 32-bit x86 and for the general function to be configured to assert if called from Windows, as a safety measure. -- This function directly executing an SVE instruction unguarded is a bit dangerous and there isn't really a way to assert that it's being done safely. It's only safe in the current use-case since its only called if the relevant support was queried from the OS.

kunalspathak replied:
> at least it's not available in the MSVC version CI is using

yes, I checked around and they haven't added that support yet, so it will not be anytime sooner

> Can we just call svcntb() on Windows instead

They do not and the guidance was to use whatever is exposed by the compiler, so we won't have it at least for new few months.

So, I think in short-term, let us do it in `asm` guarded with if CPU-capability supports SVE.

tannergooding replied:
The consideration is the msvc compiler does not support asm

kunalspathak replied:
ah ok. Yes, we should probably move this to the `.asm` file so it can be used both by the runtime and jit (`getMaxTVectorByteLength()`

tannergooding replied:
Does `MASM` support SVE instructions yet? It may be in the same boat where if MSVC doesn't support it then it may not either.

jkotas replied:
If it does not support it, you can always emit the instruction as hex code.

SwapnilGaikwad replied:
Using ACLE on Linux and return a hardcode length (128-bits) on Windows.

tannergooding replied:
> Return a hardcoded vector length of 16-bytes until we find a suitable mechanism to retrieve it.

Just noting that we won't be able to actually ship in November in such a state. We will require some form of actual query of the hardware size from the OS to avoid any issues if its run on hardware that has larger length.

It might be overall simpler to just get such a fallback setup now so we don't need to worry about it, risk forgetting it about it, or anything along those lines.

a74nh replied:
>It might be overall simpler to just get such a fallback setup now 

Agreed, but AIUI, for windows today:
* Can't use inline asm, including encoding in hex
* no OS API available
* no SVE ACLE available

I'm not aware of any other method.

kunalspathak replied:
I think the solution is what @jkotas mentioned of emitting hex code or just write the assembly `cntb` in the `.asm` and `.S` file - https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/arm64/asmhelpers.asm.

tannergooding replied:
It looks like MASM v14.4 supports things and we can just define a function that does:
```asm
rdvl    x0, #1
ret
```

SwapnilGaikwad replied:
Updated the PR to test how the ci handles use of `rdvl`.

SwapnilGaikwad committed:
Use ACLE to detect vector length on Linux and hardcode on Windows

SwapnilGaikwad commented:
Updated the PR to do the following
**On Linux**
- Disable SVE when the user specifies vector length that's smaller than the available/OS vector length.
- Avoid inline assembly and use ACLE (`svcntb()`) to retrieve the vector length.
    - Use of `svcntb()` is available from clang-16. On older versions of clang, such as clang-14, including `arm_sve.h` errors out with `SVE support not enabled` on non-SVE systems.

**On Windows**
- Return a hardcoded vector length of 16-bytes until we find a suitable mechanism to retrieve it.
- Added a TODO to note the current limitation.

a74nh commented:
> * Use of `svcntb()` is available from clang-16. On older versions of clang, such as clang-14, including `arm_sve.h` errors out with `SVE support not enabled` on non-SVE systems.

This is because the `__attribute__` `sve` is only supported from clang-16.

My only concern here is that this is pushing the minimum supported version of clang to clang-16. I'm not sure what the constraints for building coreclr are.

It would be the same if replaced with inline asm.

Alternative would be to write the asm in hex as suggested above.

SwapnilGaikwad committed:
Use rdvl instead of cntb to count vector length

SwapnilGaikwad committed:
Add inline assembly suitable for MSVC

SwapnilGaikwad commented:
Converted the PR to draft until the build issues on non-linux systems are solved.

SwapnilGaikwad committed:
Add MSCV compatible assembly helper instead of inline assembly

tannergooding commented on src/coreclr/vm/arm64/asmhelpers.S:
```diff
@@ -23,6 +23,14 @@ LEAF_ENTRY GetDataCacheZeroIDReg, _TEXT
     ret     lr
 LEAF_END GetDataCacheZeroIDReg, _TEXT
 
+#if defined(_MSC_VER)
```
Is this needed here and the same for the ifdef in the other file? I thought that `asmhelpers.S` was exclusively used by non-MSVC and `asmhelpers.asm` was exclusively used by MSVC (or rather MASM) and so there wasn't a need for such ifdefs (and why we don't have them elsewhere in these files)

kunalspathak replied:
yes, I don't think `_MSC_VER` is needed.

kunalspathak replied:
I know the call to this method is guarded under `if (SVE)`, I am actually wondering the implications of calling it by someone without having this check? Is there a way to full proof this to return 0 or something if ran on non-sve arm64 machines?

tannergooding replied:
Not really, Arm64 doesn't have a user-mode way of checking CPU feature support

kunalspathak replied:
So basically, if someone calls this method unintentionally, on non-sve arm64, they would just hit "Internal error", which is a fair thing to be.

SwapnilGaikwad replied:
Ooh, I didn't know which one was used by MSVC and non-MSVC. In this case, we need to remove the non-MSVC bits from .S file because we need the "sve" attribute to compile it.

kunalspathak commented on src/coreclr/vm/codeman.h:
```diff
@@ -1912,6 +1916,16 @@ protected :
         return m_CPUCompileFlags;
     }
 
+#if defined(TARGET_ARM64) && !defined(_MSC_VER)
```
no need of this anymore?

SwapnilGaikwad replied:
We need an "sve" attribute for GNUC builds. Not sure if `#if defined(TARGET_ARM64) && defined(__GNUC__)` is enough here.

SwapnilGaikwad replied:
I can only test on Arm64 host with Linux base systems at the moment so have to spam this CI for other build combinations.

kunalspathak replied:
You do not need this definition because you will be using the one that is defined in asmhelpers.* files.

SwapnilGaikwad replied:
Apologies for the delay in responding. It doesn't compile after removing the definition of `GetSveLengthFromOS` from here.
The issue here is that on Linux, if we rely on asmhelpers.* files, we get compilation error `error: instruction requires: sve or sme`. 
We avoid this error by adding the `__attribute__((target("sve")))`. However, this is attribute not defined for MSVC so we use asmhelpers.* files to define `GetSveLengthFromOS` there.

jkotas replied:
What does `__attribute__((target("sve")))` traslate to in the .S file? I think you should be able to make the .S file compile by adding assembly equivalent of `__attribute__((target("sve")))` to it.

jkotas replied:
Have you looked into using  `.arch_extension sve` in the .S file to get the SVE instruction to compile? It would look better if the .asm and .S files are symmetric.

kunalspathak commented on src/coreclr/vm/codeman.h:
```diff
@@ -69,6 +69,10 @@ Module Name:
 #include "gcinfo.h"
 #include "eexcp.h"
 
+#if defined(TARGET_ARM64) && defined(_MSC_VER)
```
```suggestion
#if defined(TARGET_ARM64)
```

SwapnilGaikwad replied:
Done

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Declare GetSveLengthFromOS on Arm64

jkotas commented on src/coreclr/vm/arm64/asmhelpers.asm:
```diff
@@ -78,6 +78,14 @@
         ret     lr
     LEAF_END
 
+#if defined(_MSC_VER)
+;; DWORD64 __stdcall GetSveLengthFromOS(void);
+    LEAF_ENTRY GetSveLengthFromOS
+        rdvl    x0, 1
+        ret     lr
+    LEAF_END
+#endif // _MSC_VER
+
```
```suggestion
;; DWORD64 __stdcall GetSveLengthFromOS(void);
    LEAF_ENTRY GetSveLengthFromOS
        rdvl    x0, 1
        ret     lr
    LEAF_END

```
The .asm file is only used with MSVC.

SwapnilGaikwad replied:
Done.

jkotas commented on src/coreclr/vm/arm64/asmhelpers.S:
```diff
@@ -23,6 +23,14 @@ LEAF_ENTRY GetDataCacheZeroIDReg, _TEXT
     ret     lr
 LEAF_END GetDataCacheZeroIDReg, _TEXT
 
+#if defined(_MSC_VER)
```
```suggestion
```
.S file is only used with gcc/clang

SwapnilGaikwad replied:
ooh cool. Thanks for confirming it. I'll move the definition from .S file as it would fail on clang with `error: instruction requires: sve or sme`.

jkotas replied:
You should be able to use `.arch_extension sve` in the .S file to get the SVE instruction to compile

SwapnilGaikwad replied:
Done.

SwapnilGaikwad committed:
Remove definition of GetSveLengthFromOS from .S file

SwapnilGaikwad committed:
Move non-windows definition of GetSveLengthFromOS to .S file

SwapnilGaikwad committed:
Merge main

jkotas commented on src/coreclr/vm/arm64/asmhelpers.S:
```diff
@@ -23,6 +23,13 @@ LEAF_ENTRY GetDataCacheZeroIDReg, _TEXT
     ret     lr
 LEAF_END GetDataCacheZeroIDReg, _TEXT
 
+// DWORD64 __stdcall GetSveLengthFromOS(void);
```
```suggestion
// uint64_t GetSveLengthFromOS(void);
```

jkotas commented on src/coreclr/vm/arm64/asmhelpers.asm:
```diff
@@ -78,6 +78,12 @@
         ret     lr
     LEAF_END
 
+;; DWORD64 __stdcall GetSveLengthFromOS(void);
```
```suggestion
;; uint64_t GetSveLengthFromOS(void);
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -69,6 +69,10 @@ Module Name:
 #include "eexcp.h"
 
+#if defined(TARGET_ARM64)
+extern "C" DWORD64 __stdcall GetSveLengthFromOS();
```
```suggestion
extern "C" uint64_t GetSveLengthFromOS();
```

jkotas reviewed (commented):
We prefer standard C/C++ types for new code.

jkotas committed:
Apply suggestions from code review

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -69,6 +69,10 @@ Module Name:
 #include "eexcp.h"
 
+#if defined(TARGET_ARM64)
+extern "C" uint64_t GetSveLengthFromOS();
```
Could you please move this to the .cpp file (e.g. next to `GetDataCacheZeroIDReg`). It does not need to be in the header file.

SwapnilGaikwad replied:
Done.

SwapnilGaikwad committed:
Move declaration of GetSveLengthFromOS to .cpp file

SwapnilGaikwad commented:
Hi @jkotas, do you have any suggestions to fix[ this build error](https://dev.azure.com/dnceng-public/public/_build/results?buildId=698281&view=logs&j=5d780cf0-7dc0-5aa0-741e-ecc17ef09a05&t=a5e632ba-8fa1-50be-a73e-ed850090cc2e&l=2620) - `unknown opcode: rdvl`, for Windows on Arm64 ?
Potentially any current places where we use a hex code for an instruction.

jkotas commented on src/coreclr/vm/arm64/asmhelpers.asm:
```diff
@@ -78,6 +78,12 @@
 
+;; uint64_t GetSveLengthFromOS(void);
+    LEAF_ENTRY GetSveLengthFromOS
+        rdvl    x0, 1
```
```suggestion
        ;; TODO: Uncomment once CI machines are updated to use MASM 14.4 or later
        ;; rdvl    x0, 1
```

SwapnilGaikwad replied:
Done.

jkotas commented:
> do you have any suggestions to fix[ this build error](https://dev.azure.com/dnceng-public/public/_build/results?buildId=698281&view=logs&j=5d780cf0-7dc0-5aa0-741e-ecc17ef09a05&t=a5e632ba-8fa1-50be-a73e-ed850090cc2e&l=2620) - unknown opcode: rdvl, for Windows on Arm64 ?

I do not think we need to be creative: https://github.com/dotnet/runtime/pull/101295/files#r1628086182

SwapnilGaikwad committed:
Disable  temporarily for Windows on Arm machines

tannergooding commented on src/coreclr/vm/arm64/asmhelpers.asm:
```diff
@@ -78,6 +78,13 @@
+;; uint64_t GetSveLengthFromOS(void);
+    LEAF_ENTRY GetSveLengthFromOS
+        ;; TODO-SVE: Uncomment once CI machines are updated to use MASM 14.4 or later
+        ;; rdvl    x0, 1
```
Should this explicitly zero `x0` until the right MASM version can be used, so that it's state isn't undefined?

SwapnilGaikwad replied:
>we should at least make it return 128 for windows/arm64

Returning hardcoded 128 for now.

kunalspathak commented:
> > do you have any suggestions to fix[ this build error](https://dev.azure.com/dnceng-public/public/_build/results?buildId=698281&view=logs&j=5d780cf0-7dc0-5aa0-741e-ecc17ef09a05&t=a5e632ba-8fa1-50be-a73e-ed850090cc2e&l=2620) - unknown opcode: rdvl, for Windows on Arm64 ?
> 
> I do not think we need to be creative: https://github.com/dotnet/runtime/pull/101295/files#r1628086182

Currently, `GetSveLengthFromOS()` will return different values each time depending on the content of `x0` for windows/arm64. Until we upgrade the masm on CI machine (not sure how frequently that happens), we should at least make it return `128` for windows/arm64.

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Hardcode GetSveLengthFromOS() to return SVE length as 128

kunalspathak reviewed (approved):
LGTM. Thanks for your contributions.
